src/nodes/envelopes.rs

```diff
@@ -0,0 +1,452 @@
+            .uses_events(true)
+    }
+
+    fn construct_processor(
+        &self,
+        _config: &Self::Configuration,
+        cx: ConstructProcessorContext,
+    ) -> impl AudioNodeProcessor {
+        let lo = self.lo.linear();
+        let hi = self.hi.linear();
+        let sustain_proportion = self.sustain_proportion.clamp(0., 1.);
+
+        let attack = ClockSeconds(self.attack);
+        let hold = ClockSeconds(self.hold);
+        let decay = ClockSeconds(self.decay);
+        let release = ClockSeconds(self.release);
+
+        AhdsrVolumeProcessor {
+            state: None,
+            current: SmoothedParam::new(lo, Default::default(), cx.stream_info.sample_rate),
+
+            lo,
+            hi,
+
+            attack_rate: (cx.stream_info.sample_rate_recip / attack.0) as f32,
+            hold_samples: ClockSamples::from_secs_f64(hold.0, cx.stream_info.sample_rate.get()),
+            decay_rate: (1. - sustain_proportion)
+                * (cx.stream_info.sample_rate_recip / decay.0) as f32,
+            sustain_proportion,
+            release_rate: sustain_proportion
+                * (cx.stream_info.sample_rate_recip / release.0) as f32,
+
+            attack,
+            hold,
+            decay,
+            release,
+
+            off_state_transition: self.trigger_mode.state_transition(),
+            on_state_transition: self.retrigger_mode.state_transition(),
+            sustain_state: self.trigger_mode.sustain_state(),
+
+            sample_rate: cx.stream_info.sample_rate,
+            sample_rate_recip: cx.stream_info.sample_rate_recip,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum AhdsrState {
+    Attack,
+    /// The inner value is the number of samples that the value has been held for.
+    Hold(i64),
+    Decay,
+    Sustain,
+    Release,
+}
+
+struct AhdsrVolumeProcessor {
+    state: Option<AhdsrState>,
+
+    /// Normalized 0..1, with 0 = lo, 1 = hi. This allows us to change `hi`/`lo`/`sustain` while the envelope
+    /// is activated.
+    current: SmoothedParam,
+
+    lo: f32,
+    hi: f32,
+
+    attack_rate: f32,
+    hold_samples: ClockSamples,
+    decay_rate: f32,
+    sustain_proportion: f32,
+    release_rate: f32,
+
+    // -- Information to reconstruct the above fields when the node params change --
+    attack: ClockSeconds,
+    hold: ClockSeconds,
+    decay: ClockSeconds,
+    release: ClockSeconds,
+
+    // -- State transitions --
+    off_state_transition: fn(AhdsrState) -> AhdsrState,
+    sustain_state: AhdsrState,
+    on_state_transition: fn(f32, Option<AhdsrState>) -> (f32, AhdsrState),
+
+    // These are stored on `proc_info` in the latest, unreleased version of Firewheel,
+    // when that is released these fields should be removed.
+    sample_rate: NonZeroU32,
+    sample_rate_recip: f64,
+}
+
+#[inline(always)]
+fn lerp(a: f32, b: f32, ratio: f32) -> f32 {
+    let ratio = ratio.clamp(0., 1.);
+    a * (1. - ratio) + b * ratio
+}
```

Comment:
A bit of a nit, but `bevy_math` has a `lerp` method on `FloatExt` which we can use here I believe. Might as well, imo.

Reply:
Good catch! I originally put this in as a PR to Firewheel (which only optionally imports bevy to derive `Component`) so it wasn't available and didn't catch that when I moved it to this repo.